### DISCUSSION TITLE
.Net docs need to stop mentioning databricks

### DISCUSSION THREAD

**Discussion by dbeavon** [dotnet/spark#935] [Q&A]
I opened a very unsatisfying support case with databricks (via azure-databricks).  **The end result is that they don't support spark-dotnet.**  

I spent a full day building a self-contained repro, and sent it thru azure support.  The azure-databricks support team interacted extensively with project managers at databricks.  However none of the "spark core" engineers were willing to spend time digging into the repro.  As soon as they saw spark-dotnet in the stack frames they refused to look any deeper at the problem.

The challenge is that their "driver daemon" is proprietary and behaves differently than OSS spark.  So if the engineers refuse to investigate, then there is no recourse.  It is like running into a brick wall (pun intended).

I had given Terry a heads-up about this case, but I don't think anyone else contacted him.  While I do believe that he has worked with databricks engineers in the past, I don't think they were ready to support any customers yet (especially indirect customers in Microsoft Azure).

I should soon have an authoritative statement about their lack of support for .Net.

@Niharikadutta @imback82 

Is there any chance we can put warnings about supportability in these docs (links below)?
https://docs.microsoft.com/en-us/dotnet/spark/what-is-apache-spark-dotnet
https://docs.microsoft.com/en-us/dotnet/spark/tutorials/databricks-deployment

It seems to me that the docs are a bit misleading given they encourage the use of databricks, but the databricks team itself isn't willing to support .Net in any capacity whatsoever.  Previously there were warnings on these pages (prior to v.1.0.0) but the warnings were all removed.  That was the point when I thought it was safe to use databricks for .Net solutions.  Given their lack of support I think it is critical to bring the warnings back again.  Without support from the databricks engineers, there is no way to investigate why a solution that works on OSS spark will not work when deployed into their proprietary "driver daemon".

**Comment by imback82** [comment]
@dbeavon I had a call with Databricks engineers earlier this week. There are two things we are going to try:
1. From the supporting team, they will try to replicate the behavior with Scala/Python to see if this is related to Spark .NET or not.
2. I will add more logging to `ThreadPool.scala` and see what's causing the issue. I plan to share the changes with you by the end of this week so that you can try with the new JAR.

> Previously there were warnings on these pages (prior to v.1.0.0) but the warnings were all removed.

@Niharikadutta / @luisquintanilla, can you answer to this quesion why the warning were removed?

**Reply by dbeavon** [reply]
@imback82 

I'm hopeful that we might be able to push forward again.  A lot of my development effort has been spent consuming databricks-specific features like their REST API and log delivery and such.  If I were to move to a different Spark PaaS, it would probably require lots of refactoring of those types of things.  I'm still trying to hit a deadline at the end of May....

I am running fine except when the number of concurrent jobs grow higher than ~10 or so.  The original goal was to keep 50 jobs running at any given time, and their durations are generally 3-5 minutes.  If I can't do that then things will run about five times slower than planned.

If you could ask a databricks engineer to attach a debugger (Intellij), and break on first-chance exceptions in their "driver daemon", that should be enough to figure this out.  I get the feeling that trying to replicate with Scala/Python will take a lot longer and be a lot less conclusive.  I talked to the azure support team (Jeeban and Tad) today and they weren't able to get traction with the "core spark" engineers so they decided they needed to close the case.

> Previously there were warnings on these pages (prior to v.1.0.0) but the warnings were all removed

I think what I said was misleading.  I should clarify that the warnings weren't actually databricks -specific.  The warnings were related to the fact that .Net for Spark was not yet RTM'ed.  When v.1.0.0 was released and they removed the warnings in the docs, they added no other warnings in their place.  I assumed this meant that somebody spoke with databricks and agreed to start offering full support.  Unfortunately I don't think any conversations took place and databricks won't support.  

I had asked databricks a year ago about their support for .Net and what they said then isn't any different than what they are saying today (no support is available).  Which puts us in an impossible position since there is no way to see their source nor run their "driver daemon" locally.

**Reply by Niharikadutta** [reply]
@dbeavon We are sorry you are running into these issues! I checked the documentation history and like you said we did have warnings about .NET for Apache Spark being preview before version 1.0.0 which were then removed and I understand how this can be misleading regarding the level of support with Databricks. I can work on updating these docs with some warnings about the supportability in Databricks.

**Reply by dbeavon** [reply]
@Niharikadutta Thanks, its up to you if you want to take my word for it on their support (... rather the lack of support).  Otherwise there is a technical support manager in azure-databricks (Chaith) who offered to send me a formal/authoritative statement on the lack of support for .Net in Azure Databricks.  He said I can share it.  I should be able to post that here soon (next week).  

I don't want to make a huge deal of this.  And I don't want to undermine everyone's SIGNIFICANT efforts to make .Net a big success (... perhaps for some that may even be in Databricks).  But I think it is important for other developers to know when they are going out on a limb with no safety net.  If I had the luxury of starting over again I think I would probably pick HDInsight.  I'll probably have to squeeze in the platform change when I start up on the next project.

**Reply by dbeavon** [reply]
@Niharikadutta 

There is an update... I now have a formal statement from Big Data support that says .Net is **unsupported on Azure Databricks**.  They created an "idea" to support .Net in the future.  The reference is DB-I-2253.  I think you need to be a direct Databricks customer to vote on it.

The following is a quote from the case closure email for case #2104270040005818. (I am allowed to share this.)

```

For your reference, here is the summary of the case. 


Symptom: The case was raised to check performance issues of .NET solution on Azure Databricks

Cause & Solution: 

1.	The dot NET code works fine on open source and gives some performance issues on Databricks environment

2.	Following are the couple documents that demonstrate how to deploy dot NET solution to databricks
  a.	https://docs.microsoft.com/en-us/dotnet/spark/tutorials/databricks-deployment
  b.	https://docs.microsoft.com/en-us/dotnet/spark/how-to-guides/databricks-deploy-methods

3.	As we found out from Databricks, only 3 languages are supported and dot NET is not one of them. Ref.: https://docs.microsoft.com/en-us/azure/databricks/languages/

4.	As dot NET is not one of the widely used language for BigData processing / Data Science, Databricks has not added it to the above list yet. Ref.: https://flatironschool.com/blog/data-science-programming-languages

5.	At point we have raised a feature request for the engineering team to add language support for dot NET on Databricks. Ref.: https://databricks.aha.io/ideas/ideas/DB-I-2253

6.	Engineering team will be prioritizing this among other feature requests and take it accordingly, due to which there is no ETA for this feature request.


```

That case summary is from Chaithanya Sarabuddi (Chaith), a Big Data Support Manager on the Azure Databricks team.

In addition to the performance issues, my support case involved timing -related issues that caused an exception to be raised:
**java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask rejected** 


Do you think we could add warnings to the pages that reference Databricks (see 2.a, and 2.b. above) ?  Another page where Databricks is referenced is https://docs.microsoft.com/en-us/dotnet/spark/what-is-apache-spark-dotnet 

The references to Databricks should indicate that they provide no support if/when a problem is encountered.  In other words using .Net is possible, but that is only at your own risk since Databricks will not support you.  This would not be such a big problem if we could otherwise investigate a problem on our own, or get some support from this community.  But that is not possible either, given that their Databricks clusters are built on a non-standard and proprietary technology ("driver daemon") which is used to host the .Net driver.  When that "driver daemon" is misbehaving there is no possible way to troubleshoot without some assistance from Databricks (which they are unwilling to provide).

**Reply by Niharikadutta** [reply]
Thanks @dbeavon for providing this statement. I will use this to update our documentation about the lack of support for .NET for Apache Spark on Databricks.

**Reply by Niharikadutta** [reply]
@imback82  please find the `microsoft-spark` jar with logging statements in `ThreadPool.scala` - [microsoft-spark-3-0_2.12-1.1.1.jar.zip](https://github.com/dotnet/spark/files/6505526/microsoft-spark-3-0_2.12-1.1.1.jar.zip)

**Reply by imback82** [reply]
Thanks @Niharikadutta! Can you share the branch with the changes as well?

@dbeavon could you try submitting your job with the JAR that @Niharikadutta shared above? This will help us debugging further.

**Reply by Niharikadutta** [reply]
Here are the changes - https://github.com/dotnet/spark/pull/936.

**Reply by dbeavon** [reply]
@imback82 @Niharikadutta  Please give me another day or so.... I had to open another related support case with databricks.  Their log4j stuff is crapping out on me.  I believe that every "spark_jar_task", when initiated, is causing the log4j level to be elevated dynamically.  This may be happening from some driver logic of theirs.

My log4j preferences for the cluster are being overridden.  The logging config changes made by cluster init scripts , *and* my own dynamic changes to the logging level are both being overridden.. 

I don't know why I keep running into issues that should have been encountered by other scala developers!  I'm guessing that the scala developers went somewhere else.  Perhaps the only users on azure databricks may be "data scientists" and they are probably isolated to using their python notebooks.  They may have lower expectations, and/or may not even care about their log4j logging.
